File: kirun-js/src/engine/function/system/math/Random.ts
import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../util/MapUtil';
import { AbstractFunction } from '../../AbstractFunction';

const VALUE = 'value';
export class Random extends AbstractFunction {
    private static readonly SIGNATURE: FunctionSignature = new FunctionSignature('Random')
        .setNamespace(Namespaces.MATH)
        .setEvents(
            new Map<string, Event>([
                Event.outputEventMapEntry(MapUtil.of(VALUE, Schema.ofDouble(VALUE))),
            ]),
        );

    public getSignature(): FunctionSignature {
        return Random.SIGNATURE;
    }
    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        return new FunctionOutput([EventResult.outputOf(new Map([[VALUE, Math.random()]]))]);
    }
}

File: kirun-py/src/kirun_py/function/system/math/random_.py
from __future__ import annotations

import random
from typing import TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.namespaces.namespaces import Namespaces

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

VALUE = 'value'


class Random(AbstractFunction):

    def __init__(self) -> None:
        super().__init__()
        self._signature = (
            FunctionSignature('Random')
            .set_namespace(Namespaces.MATH)
            .set_events(
                dict([Event.output_event_map_entry({VALUE: Schema.of_double(VALUE)})])
            )
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        return FunctionOutput([
            EventResult.output_of({VALUE: random.random()})
        ])
